addon/ext.js

```diff
@@ -331,6 +332,23 @@ JSONSerializer.reopen({
       owner.register(containerKey, transformClass);
     }
     return owner.lookup(containerKey);
+  },
+
+  // We need to override this to handle polymorphic with a typeKey function
+  applyTransforms(typeClass, data) {
+    let attributes = get(typeClass, 'attributes');
+
+    typeClass.eachTransformedAttribute((key, typeClass) => {
+      if (data[key] === undefined) {
+        return;
+      }
+
+      let transform = this.transformFor(typeClass);
+      let transformMeta = attributes.get(key);
+      data[key] = transform.deserialize(data[key], transformMeta.options, data);
```

Comment:
The only thing added compared to ember is the last argument which is the data of the parent